src/Lykke.Service.FixGateway.Services/AssetsListRequestHandler.cs

```diff
@@ -0,0 +1,127 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Common.Log;
+using JetBrains.Annotations;
+using Lykke.Service.Assets.Client;
+using Lykke.Service.Assets.Client.Models;
+using Lykke.Service.FixGateway.Core.Services;
+using QuickFix.Fields;
+using QuickFix.FIX44;
+using ILog = Common.Log.ILog;
+
+namespace Lykke.Service.FixGateway.Services
+{
+    [UsedImplicitly]
+    public sealed class AssetsListRequestHandler : IRequestHandler<SecurityListRequest>
+    {
+        private readonly IAssetsServiceWithCache _assetsServiceWithCache;
+        private readonly SessionState _sessionState;
+        private readonly IFixMessagesSender _fixMessagesSender;
+        private readonly ILog _log;
+        private readonly TimeSpan _defaultRequestTimeout = TimeSpan.FromSeconds(30);
+        private readonly CancellationTokenSource _tokenSource;
+
+
+        public AssetsListRequestHandler(IAssetsServiceWithCache assetsServiceWithCache, SessionState sessionState, IFixMessagesSender fixMessagesSender, ILog log)
+        {
+            _assetsServiceWithCache = assetsServiceWithCache;
+            _sessionState = sessionState;
+            _fixMessagesSender = fixMessagesSender;
+            _log = log.CreateComponentScope(nameof(AssetsListRequestHandler));
+            _tokenSource = new CancellationTokenSource();
+        }
+
+        public void Handle(SecurityListRequest request)
+        {
+            Task.Run(() => HandleRequestAsync(request), _tokenSource.Token); //TODO Review me!
+        }
+
+        private async Task HandleRequestAsync(SecurityListRequest request)
+        {
+            SecurityList response;
+            try
+            {
+                if (!ValidateRequest(request))
+                {
+                    return;
+                }
+
+
+                using (var cts = new CancellationTokenSource(_defaultRequestTimeout))
+                using (var cts2 = CancellationTokenSource.CreateLinkedTokenSource(cts.Token, _tokenSource.Token))
+                {
+                    var assetPairs = await _assetsServiceWithCache.GetAllAssetPairsAsync(cts2.Token);
+                    response = GetSuccessfulResponse(request, assetPairs);
+                    Send(response);
+                }
+            }
+            catch (Exception ex)
+            {
+                await _log.WriteWarningAsync(nameof(HandleRequestAsync), "Unable to receive asset information", "", ex);
+                response = GetFailedResponse(request, SecurityRequestResult.INSTRUMENT_DATA_TEMPORARILY_UNAVAILABLE);
+                Send(response);
```

Comment:
please call Send after try block and remove dublicate code + remove logic from catch block